packages/flutter/lib/src/services/spell_check.dart

```diff
@@ -0,0 +1,214 @@
+  ///
+  /// See also:
+  ///
+  ///  * [SuggestionSpan], the ranges of misspelled text and corresponding
+  ///    replacement suggestions.
+  final List<SuggestionSpan> suggestionSpans;
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) {
+        return true;
+    }
+
+    return other is SpellCheckResults &&
+        other.spellCheckedText == spellCheckedText &&
+        listEquals<SuggestionSpan>(other.suggestionSpans, suggestionSpans);
+  }
+
+  @override
+  int get hashCode => Object.hash(spellCheckedText, Object.hashAll(suggestionSpans));
+}
+
+/// Determines how spell check results are received for text input.
+abstract class SpellCheckService {
+  /// Facilitates a spell check request.
+  ///
+  /// Returns a [Future] that resolves with a [List] of [SuggestionSpan]s for
+  /// all misspelled words in [text] for the given [locale].
+  Future<List<SuggestionSpan>?> fetchSpellCheckSuggestions(
+    Locale locale, String text,
+  );
+}
+
+/// The service used by default to fetch spell check results for text input.
+///
+/// Any widget may use this service to spell check text by calling
+/// `fetchSpellCheckSuggestions(locale, text)` with an instance of this class.
+/// This is currently only supported by Android.
+///
+/// See also:
+///
+///  * [SpellCheckService], the service that this implements and may be
+///    overriden for use by [EditableText].
+///  * [EditableText], which may use this service to fetch results.
+class DefaultSpellCheckService implements SpellCheckService {
+  /// Creates service to spell check text input by default via communcication
+  /// over the spell check [MethodChannel].
+  DefaultSpellCheckService() {
+    spellCheckChannel = SystemChannels.spellCheck;
+  }
+
+  /// The last recieved [SuggestionSpan]s from the shell side.
+  List<SuggestionSpan>? lastSavedSpans;
+
+  /// The text corresponding to the [lastSavedSpans].
+  String? lastSavedText;
+
+  /// The channel used to communicate with the shell side to complete spell
+  /// check requests.
+  late MethodChannel spellCheckChannel;
+
+  /// Merges two lists of spell check [SuggestionSpan]s.
+  ///
+  /// Used in cases where the text has not changed, but the spell check results
+  /// received from the shell side have. This case is caused by IMEs (GBoard,
+  /// for instance) that ignore the composing region when spell checking text.
+  static List<SuggestionSpan> mergeResults(
+      List<SuggestionSpan> oldResults, List<SuggestionSpan> newResults) {
+    final List<SuggestionSpan> mergedResults = <SuggestionSpan>[];
+
+    SuggestionSpan oldSpan;
+    SuggestionSpan newSpan;
+    int oldSpanPointer = 0;
+    int newSpanPointer = 0;
+
+    while (oldSpanPointer < oldResults.length &&
+        newSpanPointer < newResults.length) {
+      oldSpan = oldResults[oldSpanPointer];
+      newSpan = newResults[newSpanPointer];
+
+      if (oldSpan.range.start == newSpan.range.start) {
```

Comment:
Out of curiosity do the old and new results always apple to the same text?

Also, is it possible for the ranges to start at the same place but end in different places?

e.g. [mispelledword] and [mispelled]word

Reply:
The old and new results always apply to the same text. It is only called if the text of both is the same. I left a comment where it is called noting that and I just updated the comment above this method to clarify that.

Ranges would not start at the same place and end in different places because that would be an error on the spell checker's part. This method is only to catch spans dropped in the composing region.

Reply:
Got it, thank you!